### ISSUE TITLE
Handling authentication with Website Agent

### ISSUE THREAD

**Issue by adampetrovic** [cantino/huginn#1497]
Hi There,

I've spent a couple of hours trying to wrap my head around how I'm meant to handle scraping pages that sit behind auth and therefore require a session cookie to load correctly.

Here's the basic flow I'm trying to achieve in Huginn

> 1. POST request to https://www.website.com/login with username / password in the payload
> 2. Retrieve the cookies set in the response headers of the call to /login
> 3. Send a GET request to https://www.website.com/feed (endpoint behind auth) using the cookies from 2
> 4. Extract elements of the page on /feed

I've had a couple of attempts to achieve this using a combination of the `Post Agent` and `Website Agent` using the following approach:

> 1. PostAgent configured to send a request to /login with username / password
> 2. PostAgent triggers an WebsiteAgent passing the body / headers payload to it.
> 3. Website Agent attempts to use that information to set the headers of the request. I tried using the `data_from_event` feature, however I have now learned that this is only really used to parse the event payload (rather than being able to use data in the payload to make a subsequent request).

Is there an approach I haven't considered that will allow me to achieve this?

**Comment by dsander** [comment]
Hi @adampetrovic,

your approach should work. The `WebsiteAgent` supports setting headers with the `headers` option key, you can access the headers received by the `PostAgent` via [Liquid formatting](https://github.com/cantino/huginn/wiki/Formatting-Events-using-Liquid):

```
{
  "expected_update_period_in_days": "2",
  "url": "http://xkcd.com",
  "type": "html",
  "mode": "on_change",
  "extract": {},
  "headers": {
    "Set-Cookie": "{{ headers.Set-Cookie }}"
  }
}
```

**Comment by materliu** [comment]
I think it should be
`{
  "expected_update_period_in_days": "2",
  "url": "http://xkcd.com",
  "type": "html",
  "mode": "on_change",
  "extract": {},
  "headers": {
    "Set-Cookie": "{{ headers.Set-Cookie }}"
  }
}`

**Comment by bobbysteel** [comment]
Has anyone actually gotten this properly working? I'm trying it now on an airline website but to no avail...

**Comment by cantino** [comment]
Does `"Cookie": "{{ headers['Set-Cookie'] }}"` work better?

**Comment by singler-rss** [comment]
@adampetrovic  can you post u  post agent ?

**Comment by hydewing** [comment]
any sample please?

I couldn't figure out on this website.

https://clients.micrositemasters.com/sign_in

Cheers

**Comment by silviuvulcan** [comment]
Anybody got a small example how to handle authentication? Currently I need to:
GET login page for a couple of hidden token fields
POST user, pass and the hidden fields to login saving a session cookie
GET (setting the cookie ) the page behind authentication.

I can already do it with curl and currently huginn is calling the sh script, but I'd like to use built in agents instead of grep-ing and sed-ing html.

**Comment by dsander** [comment]
I uploaded an example Scenario which handle the authentication to Huginn itself: http://huginnio.herokuapp.com/scenarios/19

**Comment by adampetrovic** [comment]
@dsander The approach you are taking doesn't work for me (at least on the website you are trying to authenticate to).

My login event passes the Set-Cookie header field with a value (redacted) as follows:
JSESSIONID=<val>; Path=/; Secure; HttpOnly, AWSELB=<val>;PATH=/;SECURE;HTTPONLY

If I try to make a request with that value as the cookie it doesn't work (website rejects me as logged out).

However if I pass the cookies as a well-formed cookie value, e.g.

JSESSIONID=<val>; AWSELB=<val>

the request works. I don't think using the 'Set-Cookie' approach is the correct long-term solution, I suspect you need to parse and extract the relevant values first.

**Comment by dsander** [comment]
@adampetrovic I agree the best solution would be to have a `extract_cookie` tag to parse the cookie string. It's not tested, but you might be able to get away with this:
```
{% assign cookies = headers['Set-Cookie'] | split: ','  %}{% for cookie in cookies %}{{cookie | split: ';' | first}};{% endfor%}
```

**Comment by pbendersky** [comment]
@dsander where would you place that template?

**Comment by dsander** [comment]
@pbendersky I think that was intended to go into a `template` key of a `WebsiteAgent` that received the cookie.

**Comment by virtadpt** [comment]
As a feature, no.  I do have something like this working to authenticate with Wallabag, though.  It's not exactly the same but I think it illustrates the back-and-forth conversation required to make it happen.

* Use a Post Agent with a payload containing the client_id, client_secret, username, and password, along with the correct grant_type ("password" in this case) to hit the server.  It must have emit_events and no_merge set "true" and output_mode should be "merge".
* In the response, use the `unescape` Liquid tag to make the response body usable.  `extract` to get the temporary access_token value.  I had to set output_type to "merge" and use a template to stuff my payload into the event.
* Use another Post Agent to contact the server.  One of the headers included the temporary access_token value.
* Style points: Trigger Agent to analyze the HTTP response code from the previous agent.
